shared-libs/transitions/src/transitions/index.js

```diff
@@ -0,0 +1,331 @@
+      });
+    })
+    .catch(err => {
+      logger.error('transitions: fetch failed for %s : %o', change.id, err);
+      return callback(err);
+    });
+};
+
+// given a collection of docs this function will:
+// - deep clone the docs, to keep the original version of them safe
+// - add _id properties to the docs that are missing them
+// - hydrate the docs and generate info docs
+// - run loaded transitions over all docs
+// - returns docs, either updated or their original version
+const processDocs = docs => {
+  if (!transitions.length) {
+    return Promise.resolve(docs);
+  }
+
+  let changes;
+  const clonedDocs = JSON.parse(JSON.stringify(docs)),
+        idsByIdx = [];
+  // keep a way to link the changed doc to the original doc
+  // there may be no reliable way to uniquely identify any of these docs
+  clonedDocs.forEach((doc, idx) => {
+    doc._id = doc._id || uuid();
+    idsByIdx[idx] = doc._id;
+  });
+
+  return lineage
+    .hydrateDocs(clonedDocs)
+    .then(hydratedDocs => {
+      changes = hydratedDocs.map(doc => ({ id: doc._id, doc: doc, seq: null }));
+      return infodoc.bulkGet(changes);
+    })
+    .then(infoDocs => {
+      changes.forEach(change => {
+        change.info = infoDocs.find(infoDoc => infoDoc.doc_id === change.id);
+      });
+      return infodoc.bulkUpdate(infoDocs);
+    })
+    .then(() => {
+      return new Promise((resolve, reject) => {
+        const operations = changes.map(change => callback => {
+          module.exports.applyTransitions(change, (err, changed) => {
+            if (!err && changed) {
+              return callback(null, change.doc);
+            }
+
+            // doc was not changed by any transition, so we return the original doc
+            // some transitions add errors to the doc but return false, so we can't assume that change.doc is clean
+            const idx = idsByIdx.findIndex(id => id === change.id);
+            return err ? callback(err) : callback(null, docs[idx]);
+          });
+        });
+        async.series(operations, (err, results) => {
+          return err ? reject(err) : resolve(results);
+        });
+      });
+    })
+    .catch(err => {
+      logger.error('Error when running transitions over docs', err);
+      return docs;
+    });
+};
+
+const applyTransitions = (change, callback) => {
+  const operations = transitions
+    .map(transition => {
+      const opts = {
+        key: transition.key,
+        change: change,
+        transition: transition.module,
+      };
+      return _.partial(applyTransition, opts, _);
+    })
+    .filter(transition => !!transition);
+
+  /*
+   * Run transitions in series and finalize. We ignore err here
+   * because it is never returned, we process it within the operation
+   * function.  All we care about are results and whether we need to
+   * save or not.
+   */
+  async.series(operations, (err, results) => finalize({ change, results }, callback));
+};
+
+
```

Comment:
`finalize` has been updated so it does not save the doc, instead it just minifies and callsback with true if the doc was updated.